### PR TITLE
Diagnose duplicated target names with different casing

### PR DESCRIPTION
**Pull request by neonichu**
This will typically lead to `import` conflicts, so we should avoid it at the model level.

rdar://117691105

### CODE REVIEW

**Comment by neonichu** [comment]
@swift-ci please test

**Comment by neonichu** [comment]
A few points still to figure out:

- the new test currently fails because we'll emit the "mirror"-style diagnostic "multiple similar targets 'Foo' appear in package 'bar' and 'foo', this may indicate that the two packages are the same and can be de-duplicated by using mirrors. if they are not duplicate consider using the `moduleAliases` parameter in manifest to provide unique names" which doesn't seem accurate here but would require changing the heuristic to fix
- we don't specifically detect that the conflict is not literal, but only when comparing case-insensitive. We need to restructure the existing code quite a bit to "remember" that bit.
- the diagnostic isn't limited to newer tools-version. This seems right to me since the problem isn't limited to newer tools-versions but maybe too aggressive since in theory there could be working packages with this problem masked in some way

**Review by MaxDesiatov — APPROVED** [review]

File: Sources/PackageGraph/PackageGraph+Loading.swift [hunk 1]
```diff
@@ -492,7 +494,8 @@ private func createResolvedPackages(
         // Establish dependencies in each target.
         for targetBuilder in packageBuilder.targets {
             // Record if we see a duplicate target.
-            foundDuplicateTarget = foundDuplicateTarget || !allTargetNames.insert(targetBuilder.target.name).inserted
+            foundDuplicateTarget = foundDuplicateTarget || !allTargetNames.insert(targetBuilder.target.name.lowercased()).inserted
```
Comment (MaxDesiatov) [inline Sources/PackageGraph/PackageGraph+Loading.swift:497]: I wonder if it warrants an extension adding a new `lowercasedName` or `normalizedName` property to `Target` or `ResolvedTarget`?

**Comment by neonichu** [comment]
Going to close since this isn't quite where I want it to be, but I also won't have time to improve it.

**Pull request closed without merging**